oak_client/src/lib.rs

```diff
@@ -22,34 +22,92 @@ pub mod proto {
 pub struct OakClient<T>
 where
-    T: micro_rpc::AsyncTransport<Error = anyhow::Error>,
+    T: micro_rpc::AsyncTransport<Error = anyhow::Error> + AsyncEvidenceProvider,
 {
     transport: T,
+    crypto_provider: SenderCryptoProvider,
 }
 
 impl<T> OakClient<T>
 where
-    T: micro_rpc::AsyncTransport<Error = anyhow::Error>,
+    T: micro_rpc::AsyncTransport<Error = anyhow::Error> + AsyncEvidenceProvider,
 {
-    pub async fn create(transport: T) -> anyhow::Result<Self> {
+    pub async fn create(mut transport: T) -> anyhow::Result<Self> {
         // TODO(#3442): Fetch enclave public key.
         // TODO(#3641): Implement client-side attestation verification.
-        Ok(Self { transport })
+        let evidence = transport
+            .get_evidence()
+            .await
+            .context("couldn't get attestation evidence")?;
+        let crypto_provider = SenderCryptoProvider::new(&evidence.encryption_public_key);
+        Ok(Self {
+            transport,
+            crypto_provider,
+        })
     }
 
     pub async fn invoke(&mut self, request_body: &[u8]) -> anyhow::Result<Vec<u8>> {
-        // TODO(#3767): Use hybrid encryption for messages.
-        self.transport
-            .invoke(request_body)
+        // Create encryptor.
+        let (serialized_encapsulated_public_key, request_encryptor) = self
+            .crypto_provider
+            .create_encryptor()
+            .context("couldn't create encryptor")?;
+
+        // Encrypt and serialize request.
+        let (request_ciphertext, response_decryptor) = request_encryptor
+            .encrypt(request_body, EMPTY_ASSOCIATED_DATA)
+            .context("couldn't encrypt request")?;
+
+        let request = HpkeRequest {
+            encrypted_message: Some(AeadEncryptedMessage {
+                ciphertext: request_ciphertext,
+                associated_data: EMPTY_ASSOCIATED_DATA.to_vec(),
+            }),
+            serialized_encapsulated_public_key: Some(serialized_encapsulated_public_key),
+        };
+        let mut serialized_request = vec![];
+        request
+            .encode(&mut serialized_request)
+            .map_err(|error| anyhow!("couldn't serialize request: {:?}", error))?;
+
+        // Send request.
+        let serialized_response = self
+            .transport
+            .invoke(&serialized_request)
             .await
-            .context("couldn't send request")
+            .map_err(|error| anyhow!("couldn't send request: {:?}", error))?;
+
+        // Deserialize and decrypt response.
+        let response = HpkeResponse::decode(serialized_response.as_ref())
+            .map_err(|error| anyhow!("couldn't deserialize response: {:?}", error))?;
+        let encrypted_message = response
+            .encrypted_message
+            .context("response doesn't contain encrypted message")?;
+        // The resulting encryptor for consequent requests is discarded because we don't send
+        // another message to the stream.
```

Comment:
Why do we need to re-create the encryptor every time? Would it not be possible to have an API that allows reusing it? Otherwise we end up with so many encryptors if we were to actually do a stream. Not for this PR, but perhaps we can follow up separately?

Reply:
I'm currently refactoring crypto API in a different PR